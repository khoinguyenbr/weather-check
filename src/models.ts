export interface CityModel {
  title: string;
  woeid: number;
}

export interface WeatherModel {
  date: Date;
  minTemp: number;
  maxTemp: number;
}

export interface IWeatherDataPayload {
  data: WeatherModel[];
  city: string;
}

export interface ICityResponse {
  title: string;
  location_type: string;
  woeid: number;
  latt_long: string;
}

export interface IWeatherResponse {
  consolidated_weather: {
    min_temp: number;
    max_temp: number;
    applicable_date: string;
  }[];
}
