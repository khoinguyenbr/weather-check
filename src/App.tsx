import React from 'react';
import { Greeting, SearchForm, WeatherList } from './components';

import './App.css';

const App: React.FC = () => {
  return (
    <React.Fragment>
      <Greeting />
      <SearchForm />
      <WeatherList />
    </React.Fragment>
  );
};

export default App;
