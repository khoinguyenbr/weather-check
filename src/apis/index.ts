import axios, { AxiosResponse } from 'axios';
import { CityModel, WeatherModel, ICityResponse, IWeatherResponse } from '../models';

function getData<T>(url: string): Promise<AxiosResponse<T>> {
  return axios.get<T>(url);
}

export const searchCity = async (city: string): Promise<CityModel[]> => {
  try {
    const { data } = await getData<ICityResponse[]>(`/api/location/search/?query=${city}`);
    return data.map((item) => ({ title: item.title, woeid: item.woeid }));
  } catch (err) {
    return [];
  }
};

export const getWeather = async (id: number): Promise<WeatherModel[]> => {
  try {
    const { data } = await getData<IWeatherResponse>(`/api/location/${id}/`);
    return data.consolidated_weather.slice(0, 5).map((item) => ({
      date: new Date(item.applicable_date),
      minTemp: item.min_temp,
      maxTemp: item.max_temp,
    }));
  } catch (err) {
    return [];
  }
};
