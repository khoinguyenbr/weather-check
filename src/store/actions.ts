import { CityModel, IWeatherDataPayload } from '../models';
import {
  START_SEARCH_CITY,
  SEARCH_CITY_SUCCESS,
  CLEAR_CITIES,
  START_REQUEST_WEATHER,
  REQUEST_WEATHER_SUCCESS,
  REQUEST_WEATHER_FAILURE,
  SearchCitySuccessAction,
  StartSearchCityAction,
  ClearCitiesAction,
  GetWeatherAction,
  GetWeatherSuccessAction,
  GetWeatherFailureAction,
} from './types';

export function startSearchCity(city: string): StartSearchCityAction {
  return { type: START_SEARCH_CITY, payload: city };
}

export function searchCitySuccess(cities: CityModel[]): SearchCitySuccessAction {
  return { type: SEARCH_CITY_SUCCESS, payload: cities };
}

export function clearCities(): ClearCitiesAction {
  return { type: CLEAR_CITIES };
}

export function getWeather(city: CityModel): GetWeatherAction {
  return { type: START_REQUEST_WEATHER, payload: city };
}

export function getWeatherSuccess(data: IWeatherDataPayload): GetWeatherSuccessAction {
  return { type: REQUEST_WEATHER_SUCCESS, payload: data };
}

export function getWeatherFailure(): GetWeatherFailureAction {
  return { type: REQUEST_WEATHER_FAILURE };
}
