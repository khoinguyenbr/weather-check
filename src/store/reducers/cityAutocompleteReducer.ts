import { CityModel } from '../../models';
import {
  CLEAR_CITIES,
  SEARCH_CITY_SUCCESS,
  START_SEARCH_CITY,
  SearchCityActionTypes,
} from '../types';

export type CityAutocompleteState = {
  isLoading: boolean;
  cities: CityModel[];
};

export const initialState: CityAutocompleteState = {
  isLoading: false,
  cities: [],
};

export default (
  state: CityAutocompleteState = initialState,
  action: SearchCityActionTypes
): CityAutocompleteState => {
  switch (action.type) {
    case START_SEARCH_CITY:
      return { ...state, isLoading: true };
    case SEARCH_CITY_SUCCESS:
      return { ...state, cities: action.payload, isLoading: false };
    case CLEAR_CITIES:
      return { ...state, cities: [], isLoading: false };
    default:
      return state;
  }
};
