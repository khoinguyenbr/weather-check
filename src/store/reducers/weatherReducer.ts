import { WeatherModel } from '../../models';
import {
  START_REQUEST_WEATHER,
  GetWeatherActionTypes,
  REQUEST_WEATHER_SUCCESS,
  REQUEST_WEATHER_FAILURE,
} from '../types';

export type WeatherState = {
  isLoading: boolean;
  city: string;
  data: WeatherModel[];
};

export const initialState: WeatherState = {
  isLoading: false,
  city: '',
  data: [],
};

export default (
  state: WeatherState = initialState,
  action: GetWeatherActionTypes
): WeatherState => {
  switch (action.type) {
    case START_REQUEST_WEATHER:
      return { ...state, isLoading: true, data: [], city: '' };
    case REQUEST_WEATHER_SUCCESS:
      const { data, city } = action.payload;
      return { ...state, isLoading: false, data, city };
    case REQUEST_WEATHER_FAILURE:
      return { ...state, isLoading: false, data: [], city: '' };
    default:
      return state;
  }
};
