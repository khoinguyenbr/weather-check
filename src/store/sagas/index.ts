import { all, fork } from 'redux-saga/effects';
import { SagaIterator } from '@redux-saga/types';

import cityAutocompleteSaga from './cityAutocompleteSaga';
import weatherSaga from './weatherSaga';

export default function* rootSaga(): SagaIterator {
  yield all([fork(cityAutocompleteSaga), fork(weatherSaga)]);
}
