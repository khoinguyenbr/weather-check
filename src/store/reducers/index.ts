import { combineReducers } from 'redux';
import cityAutocompleteReducer from './cityAutocompleteReducer';
import weatherReducer from './weatherReducer';

const rootReducer = combineReducers({
  weather: weatherReducer,
  cityAutocomplete: cityAutocompleteReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
