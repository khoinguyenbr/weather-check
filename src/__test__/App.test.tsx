/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent, waitFor } from '@testing-library/react';
import createSagaMiddleware from 'redux-saga';
import configureStore from 'redux-mock-store';
import { RootState } from '../store/reducers';
import App from '../App';
import rootSaga from '../store/sagas';
import { searchCity, getWeather } from '../apis';
import { CityModel, WeatherModel } from '../models';

jest.mock('../apis');
const mockSearchCity = searchCity as jest.Mock<Promise<CityModel[]>>;
const mockGetWeather = getWeather as jest.Mock<Promise<WeatherModel[]>>;

const sagaMiddleware = createSagaMiddleware();
const mockStore = configureStore([sagaMiddleware]);

const initialState: RootState = {
  weather: {
    isLoading: false,
    data: [],
    city: '',
  },
  cityAutocomplete: {
    isLoading: false,
    cities: [],
  },
};

const cities: CityModel[] = [
  { title: 'Ho Chi Minh', woeid: 1321 },
  { title: 'Ha Noi', woeid: 3321 },
];

const weatherData: WeatherModel[] = [
  { date: new Date('2011-01-01'), maxTemp: 38, minTemp: 28 },
  { date: new Date('2011-01-02'), maxTemp: 38, minTemp: 28 },
  { date: new Date('2011-01-03'), maxTemp: 39, minTemp: 27 },
  { date: new Date('2011-01-04'), maxTemp: 38, minTemp: 28 },
  { date: new Date('2011-01-05'), maxTemp: 38, minTemp: 28 },
];

const stateWithSuggestion = {
  ...initialState,
  cityAutocomplete: {
    ...initialState.cityAutocomplete,
    cities,
  },
};

const setUp = (state: RootState, isTestSaga = false) => {
  const store = mockStore(state);
  if (isTestSaga) {
    sagaMiddleware.run(rootSaga);
  }
  return {
    ...render(
      <Provider store={store}>
        <App />
      </Provider>
    ),
    store,
  };
};

describe('Test correct UI', () => {
  test('renders init app', () => {
    const { container } = setUp(initialState);
    expect(container).toMatchSnapshot();
  });

  test('renders app when searching city', () => {
    const { container } = setUp({
      ...initialState,
      cityAutocomplete: {
        ...initialState.cityAutocomplete,
        isLoading: true,
      },
    });
    expect(container).toMatchSnapshot();
  });

  test('renders app with city autocomplete', () => {
    const { container } = setUp(stateWithSuggestion);
    expect(container).toMatchSnapshot();
  });

  test('renders app when loading forecast', () => {
    const { container } = setUp({
      ...initialState,
      weather: {
        ...initialState.weather,
        isLoading: true,
      },
    });
    expect(container).toMatchSnapshot();
  });

  test('renders app with city forecast', () => {
    const { container } = setUp({
      ...initialState,
      weather: {
        ...initialState.weather,
        data: weatherData,
        city: 'Ho Chi Minh',
      },
    });
    expect(container).toMatchSnapshot();
  });
});

describe('Test action behavior', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  test('should do nothing when submit form', () => {
    const { getByTestId, store } = setUp(initialState);
    const form = getByTestId('search-form');
    fireEvent.submit(form);
    const actions = store.getActions();
    expect(actions).toEqual([]);
  });

  test('should update input value', () => {
    const { getByPlaceholderText, store } = setUp(initialState, true);
    const input = getByPlaceholderText('City');
    fireEvent.change(input, { target: { value: 'a' } });

    const actions = store.getActions();
    expect(actions).toEqual([]);
  });

  test('should update input value and search city if value has more than 3 characters', async () => {
    mockSearchCity.mockImplementation(() => Promise.resolve(cities));
    const { getByPlaceholderText, store } = setUp(initialState, true);
    const input = getByPlaceholderText('City');
    fireEvent.change(input, { target: { value: 'abc' } });

    await waitFor(() => {});
    const actions = store.getActions();
    expect(actions).toEqual([
      {
        payload: 'abc',
        type: 'START_SEARCH_CITY',
      },
    ]);
  });

  test('should update input value and search city, but failed then display empty', async () => {
    mockSearchCity.mockImplementation(() => Promise.reject(new Error('reject')));
    const { getByPlaceholderText, store } = setUp(initialState, true);
    const input = getByPlaceholderText('City');
    fireEvent.change(input, { target: { value: 'abc' } });

    await waitFor(() => {});
    const actions = store.getActions();
    expect(actions).toEqual([
      {
        payload: 'abc',
        type: 'START_SEARCH_CITY',
      },
    ]);
  });

  test('should update input value and clear autocomple suggestion if value is empty', async () => {
    mockSearchCity.mockImplementation(() => Promise.resolve(cities));
    const { getByPlaceholderText, store } = setUp(stateWithSuggestion, true);
    const input = getByPlaceholderText('City');
    fireEvent.change(input, { target: { value: 'weq' } });
    fireEvent.change(input, { target: { value: '' } });

    await waitFor(() => {});
    const actions = store.getActions();
    expect(actions).toEqual([
      {
        payload: 'weq',
        type: 'START_SEARCH_CITY',
      },
      {
        type: 'CLEAR_CITIES',
      },
    ]);
  });

  test('should select city when press Enter', async () => {
    mockGetWeather.mockImplementation(() => Promise.resolve(weatherData));
    const { getByPlaceholderText, store } = setUp(stateWithSuggestion, true);
    const input = getByPlaceholderText('City');
    fireEvent.keyUp(input, { key: 'Enter', code: 'Enter' });

    await waitFor(() => {});
    const actions = store.getActions();
    expect(actions).toEqual([
      {
        payload: { title: 'Ho Chi Minh', woeid: 1321 },
        type: 'START_REQUEST_WEATHER',
      },
      {
        type: 'CLEAR_CITIES',
      },
      {
        payload: { data: weatherData, city: 'Ho Chi Minh' },
        type: 'REQUEST_WEATHER_SUCCESS',
      },
    ]);
  });

  test('should not select city when press Enter without suggestion', () => {
    const { getByPlaceholderText, store } = setUp(initialState);
    const input = getByPlaceholderText('City');
    fireEvent.keyUp(input, { key: 'Enter', code: 'Enter' });

    const actions = store.getActions();
    expect(actions).toEqual([]);
  });

  test('should select city and get weather by city when click on suggestion item', async () => {
    mockGetWeather.mockImplementation(() => Promise.resolve(weatherData));
    const { getByTestId, store } = setUp(stateWithSuggestion, true);
    const cityElement = getByTestId(1321);
    fireEvent.click(cityElement);

    await waitFor(() => {});
    const actions = store.getActions();
    expect(actions).toEqual([
      {
        payload: { title: 'Ho Chi Minh', woeid: 1321 },
        type: 'START_REQUEST_WEATHER',
      },
      {
        type: 'CLEAR_CITIES',
      },
      {
        payload: { data: weatherData, city: 'Ho Chi Minh' },
        type: 'REQUEST_WEATHER_SUCCESS',
      },
    ]);
  });

  test('should select city and get weather by city but failed then display empty', async () => {
    mockGetWeather.mockImplementation(() => Promise.reject(new Error('reject')));
    const { getByTestId, store } = setUp(stateWithSuggestion, true);
    const cityElement = getByTestId(1321);
    fireEvent.click(cityElement);

    await waitFor(() => {});
    const actions = store.getActions();
    expect(actions).toEqual([
      {
        payload: { title: 'Ho Chi Minh', woeid: 1321 },
        type: 'START_REQUEST_WEATHER',
      },
      {
        type: 'CLEAR_CITIES',
      },
      {
        type: 'REQUEST_WEATHER_FAILURE',
      },
    ]);
  });
});
