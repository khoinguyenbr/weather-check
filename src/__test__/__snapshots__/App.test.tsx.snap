// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test correct UI renders app when loading forecast 1`] = `
<div>
  <div
    class="jumbotron"
  >
    <h1
      class="display-4"
    >
      Weather Demo
    </h1>
    <p>
      This is a simple weather information app
    </p>
  </div>
  <div
    class="container"
  >
    <form
      data-testid="search-form"
    >
      <div
        class="form-group"
      >
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <div
              class="input-group-text"
            >
              <img
                src="search.svg"
              />
            </div>
          </div>
          <input
            class="form-control"
            id="searchInput"
            placeholder="City"
            type="text"
            value=""
          />
          <div
            class="input-group-append"
          >
            <div
              class="input-group-text"
            >
              <span
                class="spinner-border spinner-border-sm text-primary"
                role="status"
              >
                <span
                  class="sr-only"
                >
                  Loading...
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`Test correct UI renders app when searching city 1`] = `
<div>
  <div
    class="jumbotron"
  >
    <h1
      class="display-4"
    >
      Weather Demo
    </h1>
    <p>
      This is a simple weather information app
    </p>
  </div>
  <div
    class="container"
  >
    <form
      data-testid="search-form"
    >
      <div
        class="form-group"
      >
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <div
              class="input-group-text"
            >
              <img
                src="search.svg"
              />
            </div>
          </div>
          <input
            class="form-control"
            id="searchInput"
            placeholder="City"
            type="text"
            value=""
          />
          <div
            class="input-group-append"
          >
            <div
              class="input-group-text"
            >
              <span
                class="spinner-border spinner-border-sm text-primary"
                role="status"
              >
                <span
                  class="sr-only"
                >
                  Loading...
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
  <div
    class="container"
  >
    <p
      class="text-center"
    >
      No Data
    </p>
  </div>
</div>
`;

exports[`Test correct UI renders app with city autocomplete 1`] = `
<div>
  <div
    class="jumbotron"
  >
    <h1
      class="display-4"
    >
      Weather Demo
    </h1>
    <p>
      This is a simple weather information app
    </p>
  </div>
  <div
    class="container"
  >
    <form
      data-testid="search-form"
    >
      <div
        class="form-group"
      >
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <div
              class="input-group-text"
            >
              <img
                src="search.svg"
              />
            </div>
          </div>
          <input
            class="form-control"
            id="searchInput"
            placeholder="City"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="autocomplete-wrapper"
      >
        <ul
          class="list-group"
        >
          <li
            class="list-group-item list-group-item-action"
            data-testid="1321"
          >
            Ho Chi Minh
          </li>
          <li
            class="list-group-item list-group-item-action"
            data-testid="3321"
          >
            Ha Noi
          </li>
        </ul>
      </div>
    </form>
  </div>
  <div
    class="container"
  >
    <p
      class="text-center"
    >
      No Data
    </p>
  </div>
</div>
`;

exports[`Test correct UI renders app with city forecast 1`] = `
<div>
  <div
    class="jumbotron"
  >
    <h1
      class="display-4"
    >
      Weather Demo
    </h1>
    <p>
      This is a simple weather information app
    </p>
  </div>
  <div
    class="container"
  >
    <form
      data-testid="search-form"
    >
      <div
        class="form-group"
      >
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <div
              class="input-group-text"
            >
              <img
                src="search.svg"
              />
            </div>
          </div>
          <input
            class="form-control"
            id="searchInput"
            placeholder="City"
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
  <div
    class="container"
  >
    <h3
      class="text-center p-5"
    >
      Ho Chi Minh
    </h3>
    <div
      class="card-wrapper"
    >
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            Saturday
            :
          </h5>
          <p
            class="cart-text"
          >
            Min: 
            28
            °C
          </p>
          <p
            class="cart-text"
          >
            Max: 
            38
            °C
          </p>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            Sunday
            :
          </h5>
          <p
            class="cart-text"
          >
            Min: 
            28
            °C
          </p>
          <p
            class="cart-text"
          >
            Max: 
            38
            °C
          </p>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            Monday
            :
          </h5>
          <p
            class="cart-text"
          >
            Min: 
            27
            °C
          </p>
          <p
            class="cart-text"
          >
            Max: 
            39
            °C
          </p>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            Tuesday
            :
          </h5>
          <p
            class="cart-text"
          >
            Min: 
            28
            °C
          </p>
          <p
            class="cart-text"
          >
            Max: 
            38
            °C
          </p>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h5
            class="card-title"
          >
            Wednesday
            :
          </h5>
          <p
            class="cart-text"
          >
            Min: 
            28
            °C
          </p>
          <p
            class="cart-text"
          >
            Max: 
            38
            °C
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test correct UI renders init app 1`] = `
<div>
  <div
    class="jumbotron"
  >
    <h1
      class="display-4"
    >
      Weather Demo
    </h1>
    <p>
      This is a simple weather information app
    </p>
  </div>
  <div
    class="container"
  >
    <form
      data-testid="search-form"
    >
      <div
        class="form-group"
      >
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <div
              class="input-group-text"
            >
              <img
                src="search.svg"
              />
            </div>
          </div>
          <input
            class="form-control"
            id="searchInput"
            placeholder="City"
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
  <div
    class="container"
  >
    <p
      class="text-center"
    >
      No Data
    </p>
  </div>
</div>
`;
